File: simple_pipe/operators/generic/resize.py
from __future__ import annotations

from simple_pipe.context.frame_meta import FrameMeta
from simple_pipe.enums import NodeRole
from simple_pipe.operators.base import Operator


class Resize(Operator):
    def __init__(self, node_id: str, build_init_params: dict | None = None) -> None:
        super().__init__(node_id, "resize", build_init_params)
        self._width = int(self.build_init_params.get("width", 640))
        self._height = int(self.build_init_params.get("height", 480))

    def node_role(self) -> NodeRole:
        return NodeRole.MID

    def configure(self, patch: dict):
        from simple_pipe.result import Result

        if "width" in patch:
            self._width = int(patch["width"])
        if "height" in patch:
            self._height = int(patch["height"])
        return Result.success()

    def handle_frame_meta(self, meta: FrameMeta) -> FrameMeta | None:
        out = meta.copy_shallow()
        out.extensions["resize"] = {"width": self._width, "height": self._height}
        return out

File: src/operators/generic/resize.cpp
#include "simple_pipe/operators/generic/resize.hpp"

namespace simple_pipe {

Resize::Resize(std::string node_id, nlohmann::json build_init_params)
    : Operator(std::move(node_id), "resize", std::move(build_init_params)) {
  if (build_init_params_.contains("width")) {
    width_ = build_init_params_["width"].get<int>();
  }
  if (build_init_params_.contains("height")) {
    height_ = build_init_params_["height"].get<int>();
  }
}

Result<void> Resize::Configure(const nlohmann::json& patch) {
  if (patch.contains("width")) {
    width_ = patch["width"].get<int>();
  }
  if (patch.contains("height")) {
    height_ = patch["height"].get<int>();
  }
  return Result<void>::success();
}

FrameMeta Resize::HandleFrameMeta(const FrameMeta& meta) {
  FrameMeta out = meta.ShallowCopy();
  int out_w = width_;
  int out_h = height_;
  if (out.buffer && !out.buffer->mat.empty()) {
    out_w = width_ > 0 ? width_ : out.buffer->mat.cols;
    out_h = height_ > 0 ? height_ : out.buffer->mat.rows;
  }
  out.extensions["resize"] = {{"width", out_w}, {"height", out_h}};
  return out;
}

}  // namespace simple_pipe
